stellargraph/mapper/sampled_link_generators.py

```diff
@@ -254,7 +278,9 @@ def get_levels(loc, lsize, samples_per_hop, walks):
         # of 2 nodes, so we are extracting 2 head nodes per edge
         batch_feats = []
         for hns in zip(*head_links):
-            node_samples = self.sampler.run(nodes=hns, n=1, n_size=self.num_samples)
+            node_samples = self._sampler(batch_num).run(
```

Comment:
Are batch numbers ever passed multiple times to `sample_features`? If not, it seems like this might be a memory leak: we end up caching `SampledBreadthFirstWalk` instances despite them not being used again.

As an alternative to having to have multiple `SampledBreadthFirstWalk` objects, what do you think about moving the `SampledBreadthFirstWalk.__init__` `seed` argument to `run`, or, adding an optional `seed` argument to `run` too (that gets mixed into the `__init__` seed)?

Reply:
Every epoch results in each batch number being reused, so if there are 100 batches per epoch, `self._samplers` will have exactly 100 samplers. 

I did consider something like what you're proposing, but in order to "mix" an optional `seed` with the init `seed` it seemed to me like we'd have to create a new `random` state purely off that mixed seed, which means every sample obtained from the same `seed` + optional `seed` uses an identical random state which felt like we're limiting the "random space" we use for sampling. Instead I wanted each `batch_num` to reuse a random state and mutate it each time. 

Hopefully I'm making sense!

Reply:
> Every epoch results in each batch number being reused, so if there are 100 batches per epoch, self._samplers will have exactly 100 samplers.

Ah; definitely worth a comment, if we go with this approach 😄 

> I did consider something like what you're proposing, but in order to "mix" an optional seed with the init seed it seemed to me like we'd have to create a new random state purely off that mixed seed, which means every sample obtained from the same seed + optional seed uses an identical random state which felt like we're limiting the "random space" we use for sampling. Instead I wanted each batch_num to reuse a random state and mutate it each time.

Ah, yeah, that makes sense. Maybe it could be mixed in to a sample from the random state (e.g. `random_object.randrange(2**64) + seed`), or even just overwrite it, and it becomes the caller's (i.e. here) responsibility to mix in and have global state.

Or maybe that hits all the same ordering/reproducibility problems...

Reply:
> Ah; definitely worth a comment, if we go with this approach 😄

I can add to the docstring for `def _sampler`?
```
For its intended use in a Keras/TF workflow, if there are N batches in an epoch, there will be N samplers created, each corresponding to a particular ``batch_num``.
```

> Ah, yeah, that makes sense. Maybe it could be mixed in to a sample from the random state (e.g. random_object.randrange(2**64) + seed), or even just overwrite it, and it becomes the caller's (i.e. here) responsibility to mix in and have global state. Or maybe that hits all the same ordering/reproducibility problems...

Yeah I think there will be multiple threads on the "caller" side accessing its random object with no guaranteed order. I guess one alternative might be to only keep a dictionary of random objects instead of sampler objects and generate new seeds to call `sampler.run` with on a single `sampler` object.

Reply:
Yes, but the granularity of interleaving can mean things can be invalidated/confused if things are happening in parallel, e.g. it might be that an operation like `D[x] = 1` could be interrupted half way through by another thread (running `D[y] = 2`) which might cause the write to be dropped or otherwise misbehave.

However, it looks like that shouldn't be the case, due to http://effbot.org/pyfaq/what-kinds-of-global-value-mutation-are-thread-safe.htm . In particular:

1. only one worker for each `batch_num` operates concurrently
2 mutations of global values are restricted to operations on the `self._samplers` (not individual samplers) because of 1
3. those mutations are writes like `self._samplers[batch_num] = ...`, which are atomic
4. that mutation is guaranteed to never be an overwrite: if the existence check in the `try` fails, it will never "stop failing" before the write in 3 (that is, no other thread will fill in that element concurrently), because of 1

Thus, assuming 1, I think this is fine.

An easier way to be sure of this would be to have a lock around the body of `_sampler`, which is probably fine, since that happens once at the start of the expensive sampling operations and typically runs fast.

Reply:
I've added the lock